### ISSUE TITLE
How to hide some text before or after tag when hiding tag?

### ISSUE THREAD

**Issue by streamstr** [ngs-doo/TemplaterExamples#25]
Hello.

Consider this template part:

"Template content before tag
Some text, **some text before tag1 [[SomeObject.SomeProp1]:hide] some text after tag1** [[SomeObject.SomeProp2]] some text after tag2
Template content after tag"

How can I achieve this: when [[SomeObject.SomeProp]:hide] tag is hidden I want result to look like:
"Template content before tag
Some text, value of SomeObject.SomeProp2 some text after tag2
Template content after tag"

In other words - how to mark some part of template as belonging to tag so it will be hidden or visible according to tag visibility?

I've found this sample: [https://github.com/ngs-doo/TemplaterExamples/tree/master/Intermediate/CollapseRegion](url) but there are whole rows removed on collapse of corresponding tag.

Thanks.

**Comment by zapov** [comment]
Yes, thats the way how it should be handled in Templater.

The "standard" way of doing this is to have both alternatives in the text and then remove one of them (like the tables are removed at the end of the collapse region example - based if there is some rows in them or not)

Alternative for your case (which is easier) would be to write that text into tag metadata and display it along the tag... eg:
[[SomeObject.SomeProp1]:surroundOrEmpty(some text before tag1/some text after tag1)]
But this way you will have to use same text style for that text. You have example of such simple plugin here: https://github.com/ngs-doo/TemplaterExamples/tree/master/Intermediate/BoolOverride

**Comment by streamstr** [comment]
@zapov , thanks, I got it. 

But I think that using second option ([[SomeObject.SomeProp1]:surroundOrEmpty(some text before tag1/some text after tag1)]) would lead to much more complex and less readable template: 
if natural text flow is A [[B]] C, then I should write [[B]:surroundOrEmpty(A/C)]. End users would be confused - imagine that this is template of legal contract - there will be many such conditional phrases or single words.

Duplicating entire text blocks only to add or remove single word also seems overhead to me. But at least it would lead to more readable templates.

**Comment by zapov** [comment]
One more thing which you could consider is pre-processing, eg for the purpose of converting more readable tags into Templater friendly tags.

Some of our customers are already doing that... I guess it is also applicable here. You could have

[[text before tag :B: text after tag]] as initial tag. Then your app could do the first pass through the template and convert it into [[B]:surround(text before tag/text after tag)] for the second pass

Alternatively instead of doing the second pass if you can just handle it without multiple processing thats also an option.

**Comment by streamstr** [comment]
Yes, this sounds good. Can you please point to any example? This pre-processing can be done by writing custom metadata handler plugin (IHandler)? Or with low level replacer plugin (ILowLevelReplacer)?

**Comment by zapov** [comment]
It would be just normal processing, but done twice.
There is an example of it here: https://github.com/ngs-doo/TemplaterExamples/tree/master/Advanced/DoubleProcessing

Point being... you would look at the tags and if you find such pattern use low level replace to write out new tag.
Since when you call replace even if you send in a value which would be matched, it won't actually be considered as a tag at that point. Thus you would finish the processing into a temp memory stream... and reprocess it again so that the new tags are picked up,

**Comment by zapov** [comment]
There is now example of how to do this via content control in https://github.com/ngs-doo/TemplaterExamples/tree/master/Beginner/Paragraphs
